sdk/storage/azure-storage-data-movement/src/main/java/com/azure/storage/data/movement/PathScanner.java

```diff
@@ -0,0 +1,69 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+package com.azure.storage.data.movement;
+
+import reactor.core.publisher.Flux;
+import reactor.core.publisher.Mono;
+import com.azure.core.util.logging.ClientLogger;
+
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.stream.BaseStream;
+
+public class PathScanner {
+    private final ClientLogger logger = new ClientLogger(PathScanner.class);
+    private final String basePath;
+    private final boolean isDir;
+
+    public PathScanner(String path) {
+        // Resolve the given path to an absolute path in case it isn't one already
+        Path pathObj = Paths.get(path);
+        basePath = pathObj.toAbsolutePath().toString();
+
+        // Check if the path exists and whether or not it's a directory; throw
+        // an error if there's nothing present or readable at the given path
+        if (Files.exists(pathObj)) {
+            isDir = Files.isDirectory(pathObj);
+        } else {
+            throw logger.logExceptionAsError(new IllegalArgumentException(String.format("No accessible item exists at the path '%s'.", basePath)));
+        }
+    }
+
+    public Flux<String> scan(boolean skipSubdirectories) {
+        // Iterate through the path if it's a directory
+        if (isDir) {
+            // Set up Mono for recursing through directories
+            Mono<Path> basePath = Mono.just(Paths.get(this.basePath));
+
+            // Return a Flux constructed from expanding path
+            return basePath.expand(path -> {
+                // Return an empty publisher when the path is not a directory
+                // (files will cause errors when being used with Files::list)
+                if (!Files.isDirectory(path)) {
+                    return Mono.empty();
+                } else {
+                    return Flux.using(() -> Files.list(path),
```

Comment:
Don't we need to use this to wrap the blocking list call? https://projectreactor.io/docs/core/release/reference/#faq.wrap-blocking

Reply:
Files::list returns a stream which gets manipulated with Flux::using, so I didn't think so; or is that still blocking?

Reply:
I think it's still blocking

Reply:
I'll check on this and mod it

Quick question though (@gapra-msft @kasobol-msft @amnguye):

The logic until now was set to throw from constructor if:
1. There was no file/dir at the path
2. The file/dir at the path was unreadable

But, this could be problematic since the checks don't guarantee that when we _do_ scan(), we'll still be able to read the file. Maybe it's better to only handle issue 1 above in constructor, and let 2 be dealt with at scan()? I've made the changes and pushed them so you can see and decide, but I can revert if it's not optimal.

Same applies to .NET, so the thought above applies there too (also, it would fix the inconsistency between the two where Java throws on unreadable file+dir while .NET only does so for files in ctor; don't know if that's a concern though.)

Reply:
I've made the same changes to .NET, I'm fixing the unit tests and then I'll push that to the factory PR for preview as well.

Reply:
As for your question 1, I think lets just check 1 in the constructor and 2 in the scan method, there's no point in duplicating logic if it's eventually gonna get hit